### ISSUE TITLE
Migrating projects does not take xcconfigs into account

### ISSUE THREAD

**Issue by keith** [lyptt/struct#3]
I attempted to `--migrate` our project using struct and end up with this error:

```
/Users/ksmiley/.gem/ruby/2.3.3/gems/struct-1.0.0/lib/migrator/migrator.rb:60:in `block in migrate': undefined method `include?' for nil:NilClass (NoMethodError)
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/struct-1.0.0/lib/migrator/migrator.rb:57:in `map'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/struct-1.0.0/lib/migrator/migrator.rb:57:in `migrate'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/struct-1.0.0/lib/binsupport/struct_bin.rb:142:in `do_migrate'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/struct-1.0.0/lib/binsupport/struct_bin.rb:38:in `block (2 levels) in run'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/slop-4.4.1/lib/slop/option.rb:60:in `ensure_call'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/slop-4.4.1/lib/slop/parser.rb:103:in `process'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/slop-4.4.1/lib/slop/parser.rb:110:in `try_process'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/slop-4.4.1/lib/slop/parser.rb:63:in `block in parse'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/slop-4.4.1/lib/slop/parser.rb:46:in `each'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/slop-4.4.1/lib/slop/parser.rb:46:in `each_with_index'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/slop-4.4.1/lib/slop/parser.rb:46:in `parse'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/slop-4.4.1/lib/slop/options.rb:71:in `parse'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/slop-4.4.1/lib/slop.rb:23:in `parse'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/struct-1.0.0/lib/binsupport/struct_bin.rb:24:in `run'
        from /Users/ksmiley/.gem/ruby/2.3.3/gems/struct-1.0.0/bin/struct:4:in `<top (required)>'
        from /Users/ksmiley/.gem/ruby/2.3.3/bin/struct:22:in `load'
        from /Users/ksmiley/.gem/ruby/2.3.3/bin/struct:22:in `<main>'
```

The problem is [here](https://github.com/lyptt/struct/blob/421aa21b32f43c00800928b6f27411e61c157eac/lib/migrator/migrator.rb#L60). This is caused because our entire project configuration is in xcconfigs, meaning we have 0 build settings in the `project.pbxproj`, which what causes this issue.

I'm not sure this is something you'll be interested in solving, but to do so I guess you would have to overlay the xcconfig settings on top of the project settings, this includes any other xcconfigs that use `include` internally. I'm happy to provide any more info about our config if you're interested.

Thanks!

**Comment by lyptt** [comment]
Hey Keith, thanks for reporting this!

That sounds like a decent solution to me, I'll work on this and get a new release out this week. Struct doesn't support including xcconfigs into generated projects, since xcconfigs become a bit redundant when using spec files, but for the migration process I think there's definitely a benefit in reading in any xcconfig settings when migrating projects.

I'll also fix up the underlying exception for empty build settings, because that definitely shouldn't happen.

**Comment by keith** [comment]
Awesome! That sounds great. On the note about xcconfigs being redundant, I can see that, but I actually think having the build configuration there, instead of in the yaml file for struct would be an advantage for a few reasons:

- It would be easier to transition to any other tool as we see fit, since our configs should be supported anywhere
- We get a level of inheritance in our configs. Currently we have configs organized something like this:

```
project.xcconfig
    <target_type>.xcconfig
    <configuration>.xcconfig
        <target_name>-<target_configuration>.xcconfig
```

- This is a format that iOS engineers are more likely (albeit definitely not guaranteed) to be familiar with over struct or any other third party tool like this
- This keeps our struct yaml files smaller and more readable for target dependency information which is arguably more valuable and likely to fluctuate more than the build settings
- They're more reusable across projects, if we have multiple framework dependencies they can all use the `framework.xcconfig` instead of us duplicating the settings in the yaml file

Thoughts?

**Comment by lyptt** [comment]
Hi Keith, that sounds reasonable actually. Supporting xcconfigs isn't that difficult, as the underling library we use [Xcodeproj](https://github.com/CocoaPods/Xcodeproj) allows you to specify xcconfig files when generating target or project configuration objects.

One possible solution would be to associate them in an all or nothing scenario:

```yaml
configurations:
  debug: configs/debug.xcconfig
targets:
  My app:
    configuration: configs/my-app.xcconfig
```

This would mean that you could either go with xcconfig for configuring targets, or use the built in struct method via profiles, overrides, etc.

This is probably the cleanest way of achieving this functionality, as mixing the two may make it difficult to reason about where exactly your configuration settings are coming from.

I'll add support for xcconfig files to the 1.2.0 schema, thanks for the suggestion!

**Comment by keith** [comment]
Awesome, definitely! Thank you!

**Comment by keith** [comment]
Related, we've been looking at [gyp](https://gyp.gsrc.io/) a lot lately for a similar tool as struct.

**Comment by lyptt** [comment]
This is now fixed in struct 1.0.1. Thanks for reporting the issue!
